lib/model-interfaces/langchain/functions/request-handler/adapters/shared/meta/llama2_chat.py

```diff
@@ -0,0 +1,64 @@
+import json
+
+from langchain.schema import AIMessage, HumanMessage
+from langchain.memory import ConversationBufferMemory
+from langchain.prompts import PromptTemplate
+
+
+Llama2ChatPrompt = """[INST] <<SYS>>
+You are an helpful assistant that provides concise answers to user questions with as little sentences as possible and at maximum 3 sentences. You do not repeat yourself. You avoid bulleted list or emojis.
+<</SYS>>
+
+{chat_history}
+
+{input} [/INST]"""
+
+Llama2ChatQAPrompt = """[INST] <<SYS>>
+Use the following conversation history and pieces of context to answer the question at the end. If you don't know the answer, just say that you don't know, don't try to make up an answer. You do not repeat yourself. You avoid bulleted list or emojis.
+<</SYS>>
+
+{chat_history}
+
+[INST] Context: {context} [/INST]
+
+{question} [/INST]"""
+
+Llama2ChatCondensedQAPrompt = """[INST]  <<SYS>>
+Given the following conversation and the question at the end, rephrase the follow up input to be a standalone question, in the same language as the follow up input. You do not repeat yourself. You avoid bulleted list or emojis.
+<</SYS>>
+
+{chat_history}
+
+{question} [/INST]"""
+
+
+Llama2ChatPromptTemplate = PromptTemplate.from_template(Llama2ChatPrompt)
+Llama2ChatQAPromptTemplate = PromptTemplate.from_template(Llama2ChatQAPrompt)
+Llama2ChatCondensedQAPromptTemplate = PromptTemplate.from_template(
+    Llama2ChatCondensedQAPrompt
+)
+
+
+class Llama2ConversationBufferMemory(ConversationBufferMemory):
+    @property
+    def buffer_as_str(self) -> str:
+        return self.get_buffer_string()
+
+    def get_buffer_string(self) -> str:
+        """modified version of https://github.com/langchain-ai/langchain/blob/bed06a4f4ab802bedb3533021da920c05a736810/libs/langchain/langchain/schema/messages.py#L14"""
+        human_message_cnt = 0
+        string_messages = []
+        for m in self.chat_memory.messages:
+            if isinstance(m, HumanMessage):
+                if human_message_cnt == 0:
+                    message = f"{m.content} [/INST]"
+                else:
+                    message = f"[INST] {m.content} [/INST]"
```

Comment:
this should be `f"<s> [INST] {m.content} [/INST]"`

Reply:
As you can see here https://github.com/facebookresearch/llama-recipes/blob/38df368a70c292db6e42e0f275139678497fd896/src/llama_recipes/inference/chat_utils.py#L45-L47 the code used by Langchain tokenizes each pair or human/assistant dialogs individually and then concatenates them at the end. Each tokenization add the `<s>` token at the beginning of the array `[1, ...]`, while the code explicitly adds the `</s>` token at the end `+ [tokenizer.eos_token_id]` resulting in `[1,...,2]`. Since we are passing a string an not an array of dialogues, the content is tokenized only once, and we need to add all the intermediate `<s>` and `</s>` tokens that would otherwise be generated by their code.